provisioner/aws-state.js

```diff
@@ -0,0 +1,292 @@
+'use strict';
+
+var Promise = require('promise');
+var lodash = require('lodash');
+var util = require('util');
+var Cache = require('../cache');
+var assert = require('assert');
+var debug = require('debug')('aws-provisioner:aws-state');
+
+
+/**
+ * Return a promise that resolves to an AWS State object.  We do this
+ * so that we can grab the state using promises but then inspect state
+ * using synchronus methods, as they're all just list processing
+ */
+function fetchState(ec2, keyPrefix) {
+  assert(ec2);
+  assert(keyPrefix);
+  var that = this;
+
+  var p = Promise.all([
+    ec2.describeInstances({
+      Filters: [{
+        Name: 'key-name',
+        Values: [keyPrefix + '*']
+      },{
+        Name: 'instance-state-name',
+        Values: ['running', 'pending']
+      }
+    ]}),
+    ec2.describeSpotInstanceRequests({
+      Filters: [{
+        Name: 'launch.key-name',
+        Values: [keyPrefix + '*']
+      }, {
+        Name: 'state',
+        Values: ['open']
+      }]
+    }),
+  ]);
+
+  p = p.then(function(res) {
+    return new AwsState(keyPrefix, _classify(ec2.regions, keyPrefix, res[0], res[1])
```

Comment:
I know it's legal in javascript to reference a function prior to it's definition but it scares me shitless...
This is another reason I create objects. To very much avoid this.. where you too easily and up referencing undefined functions... This is actually the reason why I write everything function as `var myFunc = function() {.... };` so it can't be referenced prior to definition.